jzaia_files/main.cpp

```diff
@@ -0,0 +1,291 @@
+ *
+ * Run Grover's algorithm from start to finish. Prepares a statevector, and
+ * repeats state selection and amplitude-amplification for sqrt(N) iterations
+ * (where N = 2^(# of non-ancilla qubits)). This implementation assumes that
+ * the oracle always picks precisly 1 state (rather than an arbitrary number).
+ *
+ * @param oracle A black-box function that acts on a created statevector
+ * @param num_qubits The number of qubits in the circuit the oracle acts
+ *        on (includes the ancilla)
+ */
+void run_experiment(void (*oracle) (StateVectorLQubitManaged<double> &),
+                    const size_t num_qubits) {
+    // Create the statevector for this circuit
+    // NOTE: qubits are numbered left-to-right order
+    StateVectorLQubitManaged<double> sv(num_qubits);
+
+    // Set up the circuit by placing into uniform superposition
+    groversSetup(sv);
+
+    // Apply (approx) sqrt(N) repetitions of state selection and amp-amp
+    const size_t nreps = static_cast<size_t>(sqrt(1llu << (num_qubits-1)));
+    for (size_t reps = nreps; reps > 0; --reps) {
+        // Apply the oracle to apply a phase of -1 to desired state
+        oracle(sv);
+        // Perform amp-amp by reflecting over |+++...+>
+        groversMirror(sv);
+    }
+
+    // Set up measurements
+    Measurements<StateVectorLQubitManaged<double>> Measurer(sv);
+    // Vector to store the most common measurement outcome
+    std::vector<bool> common_result(num_qubits - 1, false);
+
+    // Perform a "measurement" by taking the expected value of this qubit over multiple runs
+    for (size_t obs_wire=0; obs_wire < num_qubits - 1; ++obs_wire) {
+        NamedObs<StateVectorLQubitManaged<double>> obs("PauliZ", {obs_wire});
+        double result = Measurer.expval(obs);
+        common_result[obs_wire] = (result < 0);
+    }
+
+    // // Commented out: Print the raw statevector output (too large and messy)
+    // std::vector<std::size_t> wires(num_qubits-1);
+    // for (size_t i = 0; i < num_qubits - 1; ++i) { wires[i] = i; }
+    // std::vector<double> probabilities = Measurer.probs(wires);
+
+    // size_t qubitString = 0;
+    // for (auto prob : probabilities) {
+    //     std::cout << qubitString << ": " << prob << std::endl;
+    //     ++qubitString;
+    // }
+
+    std::cout << "Measured results (most common with expval): " << common_result << std::endl;
+}
+
+
+int main(void) {
+    using std::chrono::high_resolution_clock;
+    using std::chrono::duration_cast;
+    using std::chrono::duration;
+    using std::chrono::milliseconds;
+
+
+    // Run experiment 1: 11010
+    std::cout << "Running Oracle 1. Expected: " << ORACLE1_EXPECTED << std::endl;
+    auto start_time = high_resolution_clock::now();
+
+    run_experiment(oracle1, ORACLE1_QUBITS);
+
+    const duration time_oracle1 = duration_cast<milliseconds>(
+        high_resolution_clock::now() - start_time);
+
+    std::cout << std::endl;
```

Comment:
Do you think you can rewrite these code blocks in terms of a loop over the three chosen oracles?

Reply:
Absolutely, the Python approach does this. I'm not sure why this didn't occur to me earlier, but I could have just done this as a loop over pairs of the oracle function pointers and the number of qubits needed for each statevector. That would have definitely been cleaner code.

Reply:
Thanks @AmintorDusko ! I think I replied to all of them thus far, let me know if you have anything else I can answer. Thanks for taking the time to read through everything and leave thoughtful feedback.

Reply:
Sorry, I was just reading back over this thread and I'm realizing I probably misinterpreted this question as "can this be done" rather than "could you do this", I'll make some changes now and push them to my fork.

Reply:
Okay, I have gone and done this in the most recent push. Actually, I also made the change from your other comment and merged the 3 oracle functions into 1.